mitodl/ol-django
# Open edX User Retirement App

Fixes https://github.com/mitodl/mitxonline/issues/697

**Description**:
This PR adds an app for Open edX User Retirement. Currently, https://github.com/openedx/tubular is the only supported way for driving user retirement. We discussed multiple approaches for driving user retirements. We could not use Tubular as a dependency as it had version compatibility issues with Python for MITxOnline and had issues with multiple packages while installing it as a dependency in edX.

Almost all of the code is imported from https://github.com/openedx/tubular. Just a management command is added and a few extra APIs are removed as we do not need those. One of the helper `_setup_all_apis_or_exit` is also removed. Originally it was being used but it had a lot of dependencies that were not needed. It had an alternative `_setup_lms_api_or_exit ` that we used.

Here is the map of the imported code from Tubular:
```
https://github.com/openedx/tubular/blob/master/tubular/edx_api.py -> src/mitol/openedx_user_retirement/edx_api.py
https://github.com/openedx/tubular/blob/master/tubular/exception.py#L70 -> src/mitol/openedx_user_retirement/exception.py
https://github.com/openedx/tubular/blob/master/tubular/scripts/helpers.py -> src/mitol/openedx_user_retirement/helpers.py
https://github.com/openedx/tubular/blob/master/tubular/scripts/retire_one_learner.py - > src/mitol/openedx_user_retirement/management/commands/retire_one_learner.py (Added the management command)
```

**Testing Guidelines:**

**_SETTING UP EDX:_** Detailed guidelines can be found [here](https://edx.readthedocs.io/projects/edx-installing-configuring-and-running/en/latest/configuration/user_retire/implementation_overview.html).

- Go to the LMS shell and add the following settings in /edx/etc/lms.yml.

```
FEATURES:
    .....
    ENABLE_BULK_USER_RETIREMENT: true
RETIREMENT_SERVICE_WORKER_USERNAME=retirement_service_worker
RETIREMENT_STATES:
- PENDING
- RETIRING_FORUMS
- FORUMS_COMPLETE
- RETIRING_ENROLLMENTS
- ENROLLMENTS_COMPLETE
- RETIRING_NOTES
- NOTES_COMPLETE
- RETIRING_PROCTORING
- PROCTORING_COMPLETE
- RETIRING_LMS_MISC
- LMS_MISC_COMPLETE
- RETIRING_LMS
- LMS_COMPLETE
- ERRORED
- ABORTED
- COMPLETE
```

- Now run the following command `./manage.py lms --settings=devstack_docker populate_retirement_states`. It will add the above retirement states. You can check these states at `/admin/user_api/retirementstate/`.
- Now create a retirement service user and OAuth Application. Copy the Client ID and Secret.
```
app_name=retirement
user_name=retirement_service_worker
./manage.py lms --settings=devstack_docker manage_user $user_name $user_name@example.com --staff --superuser
./manage.py lms --settings=devstack_docker create_dot_application $app_name $user_name
```
- You will also need to configure `edx_notes_api` if not already. You will get 500 if edx_notes_api is not set up. Here are the minimal steps:
  - Go to the devstack and do `make dev.pull.edx_notes_api`
  - `make dev.provision.edx_notes_api` and `make dev.up.lms+edx_notes_api`
  - Now add an OAuth Application in LMS Django admin at `/admin/oauth2_provider/application/` like the below SS
  
<img width="1106" alt="Screen Shot 2023-01-10 at 11 16 42 AM" src="https://user-images.githubusercontent.com/52656433/211476035-caf0e099-1f3d-442c-b117-123d35edbfd5.png">



**_SETTING UP MITXONLINE/ANY OTHER APP:_**

- Add the following settings in `.env`
```
MITOL_OPENEDX_USER_RETIREMENT_CLIENT_ID=<Retirement OAuth Application Client ID copied above>
MITOL_OPENEDX_USER_RETIREMENT_CLIENT_SECRET=<Retirement OAuth Application Client Secret copied above>
OPENEDX_API_BASE_URL=http://host.docker.internal:18000 (Should be there already)
```


**_INSTALLING AND TESTING:_**

- Clone https://github.com/mitodl/ol-django. Checkout `asad/697-user-retirement`
- Create a virtual environment for ol-django. Install all the requirements.
- Run `./pants package :: --no-pantsd`
- Install .tar.gz or .whl file created for openedx_user_retirement in MITxOnline.
- Add the following settings if the MITxOnline PR is not up. Will update this description if PR is up.
```
from mitol.openedx_user_retirement.settings.openedx_user_retirement import *
INSTALLED_APPS = [
....
"mitol.openedx_user_retirement.apps.OpenedxUserRetirementApp"
]
```
- Create a new user to retire.
- Move the user to the retirement pipeline if not already. You can check retirement statuses at `/admin/user_api/userretirementstatus/`. If the user is not there then you can do one of the following:
  - Go to **_lms shell_** and run `./manage.py lms retire_user --username <username> --user_email <user_email>`. It will move the user to the retirement Pipeline in edX.
  - Add a flag `--initiate_retirement` while running `retire_one_learner` command mentioned in the next step.
- Go to the MITxOnline shell and Run `./manage.py retire_one_learner --user=<username> --initiate_retirement`. **[NOTE]** DO NOT ADD `--initiate_retirement` if the user is already moved to the retirement pipeline.
- User should be retired if everything goes fine. You can verify in `/admin/user_api/userretirementrequest/` and `/admin/user_api/userretirementstatus/`
